xarray/core/rolling_exp.py

```diff
@@ -0,0 +1,36 @@
+
+from pandas.core.window import _get_center_of_mass
+
+from .computation import apply_ufunc
+
+
+def _get_alpha(com=None, span=None, halflife=None, alpha=None):
+    # pandas defines in terms of com (converting to alpha in the algo)
+    # so use its function to get a com and then convert to alpha
+
+    com = _get_center_of_mass(com, span, halflife, alpha)
+    return 1 / (1 + com)
+
+
+class RollingExp(object):
+    _attributes = ['alpha', 'dim']
+
+    def __init__(self, obj, windows, window_type='span'):
+        self.obj = obj
+        dim, window = next(iter(windows.items()))
+        self.dim = dim
+        self.alpha = _get_alpha(**{window_type: window})
+
+
+class DataArrayRollingExp(RollingExp):
+    def mean(self):
+        from numbagg.moving import rolling_exp_nanmean
+
+        da = apply_ufunc(
+            rolling_exp_nanmean,
+            self.obj,
+            input_core_dims=[[self.dim]],
+            output_core_dims=[[self.dim]],
+            kwargs=dict(window=self.alpha),
+        )
+        return da.transpose(*self.obj.dims)
```

Comment:
As I noted in https://github.com/pydata/xarray/pull/2674, I think we could write a generic `Dataset`/`DataArray`/`Variable` version of this with `reduce`:
```python
def rolling_exp_nanmean(array, *, axis, window):
    import numbagg
    if axis == ():
        return 1.0 * array  # cast to float
    else:
        return numbagg.rolling_exp_nanmean(array, axis=axis, window=window)

class RollingExp:
    ...
    def mean(self):
        return self.obj.reduce(rolling_exp_nanmean, dim=self.dim, window=self.alpha)
```

Reply:
Amazing! Thank you

Reply:
Should we promote this part to the function (in this case, to numbagg)?

```
def rolling_exp_nanmean(array, *, axis, window):
    import numbagg
    if axis == ():
        return 1.0 * array  # cast to float
    else:
        return numbagg.rolling_exp_nanmean(array, axis=axis, window=window)
```